packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -1438,6 +1451,77 @@ export class SnapController extends BaseController<
     return encodeAuxiliaryFile(value, encoding);
   }
 
+  /**
+   * Processes dynamic permission request.
+   *
+   * @param snapId - The id of the Snap requesting permissions.
+   * @param permissions - Structure with requested permissions.
+   * @returns Result of permission grant request.
+   */
+  async requestPermissions(
+    snapId: SnapId,
+    permissions: SnapPermissions,
+  ): Promise<SubjectPermissions<PermissionConstraint>> {
+    log(`Granting dynamic permissions to the Snap: ${snapId}`);
+    const snapsState = this.state.snaps;
+    const snap = snapsState[snapId];
+    const { dynamicPermissions } = snap; // Use to validate
+
+    if (!dynamicPermissions) {
+      throw new Error('This Snap have no dynamic permissions specified.');
+    }
+
+    const requestedPermissionsKeys = Object.keys(permissions);
+    const dynamicPermissionKeys = Object.keys(dynamicPermissions);
+
+    // Ensure that requested dynamic permissions are allowed
+    if (
+      !requestedPermissionsKeys.every((key) =>
+        dynamicPermissionKeys.includes(key),
+      )
+    ) {
+      throw new Error(
+        'Permissions requested do not match the allowed dynamic permissions specified in the manifest.',
+      );
+    }
+
+    const pendingApproval = this.#createApproval({
```

Comment:
I'm wondering if we couldn't just use `PermissionController:requestPermissions` instead of a custom approval for this? 🤔

Reply:
Refactored to use hook to the PermissionController: https://github.com/MetaMask/metamask-extension/blob/e956826aabb2b17508d66690170fca5a9c99daa0/app/scripts/metamask-controller.js#L4839